STM/src/main/java/org/jboss/stm/internal/optimistic/OptimisticLockRecord.java

```diff
@@ -135,7 +135,7 @@ public int topLevelCommit ()
     {       
         boolean stateOK = checkState();
         
-        if (stateOK)
+        if (!stateOK)
```

Comment:
If you have a test case that is triggering this behaviour I guessed you would have seen the rollback anyway. No need to double check if not.